kbcstorage/base.py

```diff
@@ -0,0 +1,139 @@
+"""
+Base classes for constructing the client.
+
+Primarily exposes a base Endpoint class which deduplicates functionality across
+various endpoints, such as tables, workspaces, jobs, etc. as described in the
+`Storage API documentation`.
+
+
+.. _Storage API documentation:
+    http://docs.keboola.apiary.io/
+"""
+from urllib.parse import urljoin
+
+import requests
+
+
+class Endpoint:
+    """
+    Base class for implementing a single endpoint related to a single entities
+    as described in the Storage API.
+
+    Attributes:
+        path (str): URL for this endpoint.
+        token (str): A key for the Storage API.
+    """
+    def __init__(self, root, extension, token):
+        """
+        Create an endpoint.
+
+        Args
+            root (str): Root url of API. eg.
+                "https://connection.keboola.com/v2/storage/"
+            extension (str): Extension of url for the endpoint. eg. "buckets"
+            token (str): A key for the Storage API. Can be found in the storage
+            console.
+        """
+        self.path = urljoin(root, extension)
+        self.token = token
+
+    def _get(self, params=[], extra_headers={}):
```

Comment:
Maybe params is the wrong variable name for that argument. From what I can see, there are two types of params that can be passed: parameters which form the query string, like the parameters for `list-bucket-events`, such as `sinceId` etc; and parameters which go in the *path* of the url, such as `workspace_id`. The params arg as it stands it stands is solely for the latter, and maybe should be renamed. Query string params can be passed as a dict to `requests.get`.

Writing this code, and trying to made the `Endpoint._get` and `Endpoint._post` methods as versatile as possible it seems that we have to essentially replicate the `requests.get` and `requests.post` functionality which is redundant. I've been thinking about suggesting some changes in #6. the `Endpoint` class is still useful, but perhaps each endpoint subclass should be constructing its own requests, as they are so straightforward and the `Endpoint` superclass should instead implement response handling and header generation.